polkadot/node/tracking-allocator/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Tracking global allocator. Calculates the peak allocation between two checkpoints.
+
+use core::alloc::{GlobalAlloc, Layout};
+use std::{
+	ptr::null_mut,
+	sync::atomic::{AtomicBool, Ordering},
+};
+
+struct TrackingAllocatorData {
+	lock: AtomicBool,
+	current: isize,
+	peak: isize,
+	limit: isize,
+	failure_handler: Option<Box<dyn Fn()>>,
+}
+
+impl TrackingAllocatorData {
+	#[inline]
+	fn lock(&self) {
+		loop {
+			// Try to acquire the lock.
+			if self
+				.lock
+				.compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
+				.is_ok()
+			{
+				break
+			}
+			// We failed to acquire the lock; wait until it's unlocked.
+			//
+			// In theory this should result in less coherency traffic as unlike `compare_exchange`
+			// it is a read-only operation, so multiple cores can execute it simultaneously
+			// without taking an exclusive lock over the cache line.
+			while self.lock.load(Ordering::Relaxed) {
+				std::hint::spin_loop();
+			}
+		}
+	}
+
+	#[inline]
+	fn unlock(&self) {
+		self.lock.store(false, Ordering::Release);
+	}
+
+	fn start_tracking(&mut self, limit: isize, failure_handler: Option<Box<dyn Fn()>>) {
+		self.lock();
+		self.current = 0;
+		self.peak = 0;
+		self.limit = limit;
+		self.failure_handler = failure_handler;
```

Comment:
Honestly, I'd better panic if it's called a second time. It's an obvious error to do so. But unwinding from the global allocator is UB, so your suggestion is better.

Reply:
Can you please add an explanatory comment above `_old_handler`? Might even be a good idea to manually call `drop(old_handler)` after `unlock`, to make it explicit *and* ensure the value lives that long. Otherwise someone in the future might be tempted to change `_old_handler` to just `_`, which would immediately drop.

Reply:
Yeah, this should have an explicit call to `core::mem::drop` for clarity.